src/wallet/coin_selection.rs

```diff
@@ -298,45 +261,52 @@ impl<D: Database> CoinSelectionAlgorithm<D> for OldestFirstCoinSelection {
                 .chain(optional_utxos.into_iter().map(|utxo| (false, utxo)))
         };
 
-        // Keep including inputs until we've got enough.
-        // Store the total input value in selected_amount and the total fee being paid in fee_amount
-        let mut selected_amount = 0;
-        let selected = utxos
-            .scan(
-                (&mut selected_amount, &mut fee_amount),
-                |(selected_amount, fee_amount), (must_use, weighted_utxo)| {
-                    if must_use || **selected_amount < amount_needed + **fee_amount {
-                        **fee_amount +=
-                            fee_rate.fee_wu(TXIN_BASE_WEIGHT + weighted_utxo.satisfaction_weight);
-                        **selected_amount += weighted_utxo.utxo.txout().value;
-
-                        log::debug!(
-                            "Selected {}, updated fee_amount = `{}`",
-                            weighted_utxo.utxo.outpoint(),
-                            fee_amount
-                        );
-
-                        Some(weighted_utxo.utxo)
-                    } else {
-                        None
-                    }
-                },
-            )
-            .collect::<Vec<_>>();
-
-        let amount_needed_with_fees = amount_needed + fee_amount;
-        if selected_amount < amount_needed_with_fees {
-            return Err(Error::InsufficientFunds {
-                needed: amount_needed_with_fees,
-                available: selected_amount,
-            });
-        }
+        select_sorted_utxos(utxos, fee_rate, amount_needed, fee_amount)
+    }
+}
 
-        Ok(CoinSelectionResult {
-            selected,
-            fee_amount,
-        })
+fn select_sorted_utxos(
```

Comment:
As far as I can tell there are now three coin selection algorithms that order UTXOs in some fashion and then accumulate from the front: Largest First, Oldest First, and Single Random Draw.

I think the new function may replicate this part of SRD: https://github.com/bitcoindevkit/bdk/pull/148/files#diff-fd2a47357bd088bed85419710e86d949d1a172691db87c978be1336a725a42afR491-R502

Perhaps all three should use one the same accumulate function after the UTXO ordering?

Reply:
Yes, that's a good point, SRD could probably be refactored to use this function